yggdrasil/src/main/scala/quasar/yggdrasil/table/Slice.scala

```diff
@@ -1883,16 +1883,96 @@ object Slice {
     }
   }
 
+  final case class ArraySliced[A](arr: Array[A], start: Int, size: Int) {
+    def head: A = arr(start)
+    def tail: ArraySliced[A] = ArraySliced(arr, start + 1, size - 1)
+  }
+
+  def fromRValuesStep(values: ArraySliced[RValue], maxRows: Int, maxColumns: Int): (Slice, ArraySliced[RValue]) = {
+    def inner(next: ArraySliced[RValue], rows: Int, cols: Int, acc: Map[ColumnRef, ArrayColumn[_]], _size: Int): (Slice, ArraySliced[RValue]) = {
+      if (next.size == 0) {
+        if (acc.isEmpty) {
+          // println("no more data to use, emitting slice with no data")
+          (Slice.empty, next)
+        } else {
+          // println("no more data to use, emitting slice with data")
+          (new Slice {
+            val size = rows
+            val columns = acc
+          }, new ArraySliced[RValue](new Array[RValue](0), 0, 0))
+        }
+      } else {
+         if (rows >= _size) {
+          if (rows >= maxRows) {
+            // println("we have too many rows, cutting slice early")
+            (new Slice {
+              val size = rows
+              val columns = acc
+            }, next)
+          } else {
+            // println("we're resizing the columns because we have too many rows")
+            val newSize = _size * 2
+            inner(next, rows, cols, acc.mapValues { _.resize(newSize) }, newSize)
+          }
+        } else {
+          val flattened = next.head.flattenWithPath
+          val newAcc = updateRefs(flattened, acc, rows, maxRows)
+          val newCols = newAcc.size
+          if (newCols > maxColumns && cols > 1) {
+            // println("we have too many columns, cutting slice early")
+            (new Slice {
+              val size = rows
+              val columns = acc
+            }, next)
+          } else {
+            // println("we're adding a CValue to a column")
+            inner(next.tail, rows + 1, newCols, newAcc, maxRows)
+          }
+        }
+      }
+    }
+    inner(values, 0, 0, Map.empty, Math.max(64, Config.smallSliceSize))
```

Comment:
can we use `Config.smallSliceSize` instead of `Math.max(64, Config.smallSliceSize)`?

and rename `Config.smallSliceSize` to `Config.smallSliceRows`?